src/object/cli_coll.c

```diff
@@ -132,17 +371,30 @@ obj_coll_prep_one(struct coll_oper_args *coa, struct dc_object *obj,
 	if (rc != 0 || (shard->do_rebuilding && !coa->coa_for_modify))
 		goto out;
 
-	/* The ranks for the pool may be not continuous, let's extend the coa_dcts array. */
-	if (unlikely(shard->do_target_rank >= coa->coa_dct_cap)) {
-		D_REALLOC_ARRAY(dct, coa->coa_dcts, coa->coa_dct_cap, shard->do_target_rank + 4);
-		if (dct == NULL)
-			D_GOTO(out, rc = -DER_NOMEM);
+	D_RWLOCK_RDLOCK(&obj->cob_lock);
+
+	D_ASSERTF(shard->do_target_rank <= obj->cob_max_rank,
+		  "Unexpected shard with rank %u > %u\n", shard->do_target_rank, obj->cob_max_rank);
+	D_ASSERTF(shard->do_target_rank >= obj->cob_min_rank,
+		  "Unexpected shard with rank %u < %u\n", shard->do_target_rank, obj->cob_min_rank);
+
+	if (coa->coa_sparse) {
+		D_RWLOCK_UNLOCK(&obj->cob_lock);
+		key = shard->do_target_rank;
+		d_iov_set(&kiov, &key, sizeof(key));
+		d_iov_set(&riov, coa, sizeof(*coa));
+		d_iov_set(&viov, NULL, 0);
+		rc = dbtree_upsert(coa->coa_tree->cst_tree_hdl, BTR_PROBE_EQ, DAOS_INTENT_UPDATE,
+				   &kiov, &riov, &viov);
+		if (rc != 0)
+			goto out;
 
-		coa->coa_dcts = dct;
-		coa->coa_dct_cap = shard->do_target_rank + 4;
+		dct = viov.iov_buf;
+	} else {
+		dct = &coa->coa_dcts[shard->do_target_rank - obj->cob_min_rank];
```

Comment:
looks possible after unlock, the obj refresh layout and the "obj->cob_min_rank" changed, in that case is it still safe to use the dct pointer to assign at below code?